core/src/main/java/org/apache/iceberg/jdbc/JdbcCatalog.java

```diff
@@ -0,0 +1,391 @@
+    this.catalogName = name == null ? "jdbc" : name;
+    String fileIOImpl = properties.get(CatalogProperties.FILE_IO_IMPL);
+    this.fileIO = fileIOImpl == null ?
+            new HadoopFileIO(hadoopConf) :
+            CatalogUtil.loadFileIO(fileIOImpl, properties, hadoopConf);
+    initializeConnection(properties);
+  }
+
+  private void initializeConnection(Map<String, String> properties) {
+    try {
+      LOG.debug("Connecting to Jdbc database {}", properties.get(CatalogProperties.HIVE_URI));
+      Properties dbProps = new Properties();
+      for (Map.Entry<String, String> prop : properties.entrySet()) {
+        if (prop.getKey().startsWith(JDBC_PARAM_PREFIX)) {
+          dbProps.put(prop.getKey().substring(JDBC_PARAM_PREFIX.length()), prop.getValue());
+        }
+      }
+      dbConnPool = new JdbcClientPool(properties.get(CatalogProperties.HIVE_URI), dbProps);
+      initializeCatalogTables();
+    } catch (SQLTimeoutException e) {
+      throw new UncheckedSQLException("Database Connection timeout!", e);
+    } catch (SQLTransientConnectionException | SQLNonTransientConnectionException e) {
+      throw new UncheckedSQLException("Database Connection failed!", e);
+    } catch (SQLWarning e) {
+      throw new UncheckedSQLException("Database connection warning!", e);
+    } catch (SQLException | InterruptedException e) {
+      throw new UncheckedSQLException("Failed to initialize Jdbc Catalog!", e);
+    }
+  }
+
+  private void initializeCatalogTables() throws InterruptedException, SQLException {
+    // need to check multiple times because some databases are using different naming standard. ex: H2db keeping
+    // table names as uppercase
+    boolean exists = false;
+    DatabaseMetaData dbMeta = dbConnPool.run(Connection::getMetaData);
+    ResultSet tables = dbMeta.getTables(null, null, SQL_TABLE_NAME, null);
+    if (tables.next()) {
+      exists = true;
+    }
+    tables.close();
+    ResultSet tablesUpper = dbMeta.getTables(null, null, SQL_TABLE_NAME.toUpperCase(), null);
+    if (tablesUpper.next()) {
+      exists = true;
+    }
+    tablesUpper.close();
+    ResultSet tablesLower = dbMeta.getTables(null, null, SQL_TABLE_NAME.toLowerCase(), null);
+    if (tablesLower.next()) {
+      exists = true;
+    }
+    tablesLower.close();
+
+    // create table if not exits
+    if (!exists) {
+      dbConnPool.run(conn -> conn.prepareStatement(SQL_TABLE_DDL).execute());
+      LOG.debug("Created table {} to store iceberg tables!", SQL_TABLE_NAME);
+    }
+  }
+
+  @Override
+  protected TableOperations newTableOps(TableIdentifier tableIdentifier) {
+    return new JdbcTableOperations(dbConnPool, fileIO, catalogName, tableIdentifier);
+  }
+
+  @Override
+  protected String defaultWarehouseLocation(TableIdentifier table) {
+    if (table.hasNamespace()) {
+      return SLASH.join(warehouseLocation, table.namespace().levels(), table.name());
```

Comment:
This will result in `toString` called on the result of `levels()`, which is a `String[]`. That isn't correct. I think you want to pass `SLASH.join(table.namespace().levels())` into this `join` instead.